<h1 style="font-weight: bold;" id="top">Titanic Survival</h1>

## 1) **Import libraries**

load the libraries to be used in the notebook\
<a href="#bottom">scroll to bottom &gt;</a>


In [200]:
import os
import pprint
from typing import Any, Dict, List, Literal, Mapping, Optional, Tuple
import altair as alt
import numpy as np
import pandas as pd
from sklearn.base import BaseEstimator, ClassifierMixin
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import (
        accuracy_score,
        precision_score,
        recall_score,
        f1_score,
        confusion_matrix,
        classification_report,
    )
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import plotly.express as px
from sklearn import tree
from sklearn import ensemble
from sklearn import svm
from sklearn import neighbors
from sklearn import naive_bayes
from sklearn import linear_model
from sklearn import neural_network
from sklearn import discriminant_analysis
from sklearn import gaussian_process

In [130]:
try:
    from google.colab import drive  # type: ignore

    drive.mount("content")
    is_running_on_colab = True
except ImportError:
    is_running_on_colab = False

## 2) **Setting notebook**


In [131]:
if not is_running_on_colab:
    alt.renderers.enable("mimetype")
else:
    alt.renderers.enable("html")

## 3) **Load the data**

<p>Load the csv file to pandas dataframe</p>


In [132]:
folder_path = None
if os.path.exists("./data/"):
    folder_path = "./data/"
elif os.path.exists("content/MyDrive/Datasets/titanic_data"):
    folder_path = "content/MyDrive/Datasets/titanic_data"
else:
    raise ValueError("Unable to find titanic train and test csv files")

train_path = os.path.join(folder_path, "train.csv")
test_path = os.path.join(folder_path, "test.csv")

In [133]:
# Set the PassengerId column as index column.
index_col = "PassengerId"

# Load the train and test dataframe.
train_df = pd.read_csv(train_path, index_col=index_col)
test_df = pd.read_csv(test_path, index_col=index_col)

## 4) **Data statistics**

<p>Analyse train and test dataframe</p>


### i) Rows in train data

<p>View train dataframe top and bottom rows</p>


In [134]:
# First five rows in train data.
train_df.head()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
PassengerId,,,,,,,,,,,
1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [135]:
# First five rows in test data.
train_df.tail()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
PassengerId,,,,,,,,,,,
887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


### ii) Rows in test data

<p>View test dataframe top and bottom rows</p>


In [136]:
# First five rows in test data.
test_df.head()

,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
PassengerId,,,,,,,,,,
892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q
893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S
894,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q
895,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S
896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S


In [137]:
# Last five rows in test data
test_df.tail()

,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
PassengerId,,,,,,,,,,
1305,3,"Spector, Mr. Woolf",male,NaN,0,0,A.5. 3236,8.0500,NaN,S
1306,1,"Oliva y Ocana, Dona. Fermina",female,39.0,0,0,PC 17758,108.9000,C105,C
1307,3,"Saether, Mr. Simon Sivertsen",male,38.5,0,0,SOTON/O.Q. 3101262,7.2500,NaN,S
1308,3,"Ware, Mr. Frederick",male,NaN,0,0,359309,8.0500,NaN,S
1309,3,"Peter, Master. Michael J",male,NaN,1,1,2668,22.3583,NaN,C


### iii) Data description

<p>Get the know the datasets</p>


#### a) Data shapes


In [138]:
# Train dataframe shape.
train_shape = train_df.shape
print(f"Train shape:\n\trows: {train_shape[0]}\n\tcolumns: {train_shape[1]}")

Train shape:
	rows: 891
	columns: 11


In [139]:
# Test dataframe shape.
test_shape = test_df.shape
print(f"Test shape:\n\trows: {test_shape[0]}\n\tcolumns: {test_shape[1]}")

Test shape:
	rows: 418
	columns: 10


#### b) Data column names


In [140]:
columns = train_df.columns.tolist()
pprint.pprint(columns)

['Survived',
 'Pclass',
 'Name',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Ticket',
 'Fare',
 'Cabin',
 'Embarked']


In [141]:
train_df.dtypes

Survived      int64
Pclass        int64
Name         object
Sex          object
Age         float64
SibSp         int64
Parch         int64
Ticket       object
Fare        float64
Cabin        object
Embarked     object
dtype: object

#### i) Converting train data frame low unique values


In [142]:
def convert_low_unique_value_to_str(data: pd.DataFrame) -> pd.DataFrame:
    """
    Convert low unique values columns to string.

    Args
    ----
    data: Dataframe to change low unique columns.

    :return: Dataframe.
    """
    # Change Survived, Pclass, SibSp and Parch to string data type for analysis.
    data_copy = data.copy()

    # Loop over the column name to turn to object
    for col in ("Survived", "Pclass", "SibSp", "Parch"):
        data_copy[col] = data_copy[col].astype(str)

    return data_copy


ana_train_data = convert_low_unique_value_to_str(train_df)
ana_train_data.dtypes

Survived     object
Pclass       object
Name         object
Sex          object
Age         float64
SibSp        object
Parch        object
Ticket       object
Fare        float64
Cabin        object
Embarked     object
dtype: object

#### c) Data information


In [143]:
# Train data information.
train_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 891 entries, 1 to 891
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Name      891 non-null    object 
 3   Sex       891 non-null    object 
 4   Age       714 non-null    float64
 5   SibSp     891 non-null    int64  
 6   Parch     891 non-null    int64  
 7   Ticket    891 non-null    object 
 8   Fare      891 non-null    float64
 9   Cabin     204 non-null    object 
 10  Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 83.5+ KB


In [144]:
# Test data information
test_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 418 entries, 892 to 1309
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pclass    418 non-null    int64  
 1   Name      418 non-null    object 
 2   Sex       418 non-null    object 
 3   Age       332 non-null    float64
 4   SibSp     418 non-null    int64  
 5   Parch     418 non-null    int64  
 6   Ticket    418 non-null    object 
 7   Fare      417 non-null    float64
 8   Cabin     91 non-null     object 
 9   Embarked  418 non-null    object 
dtypes: float64(2), int64(3), object(5)
memory usage: 35.9+ KB


#### d) Describing the data


In [145]:
# Train dataframe description
train_df.describe().T

,count,mean,std,min,25%,50%,75%,max
Survived,891.0,0.383838,0.486592,0.00,0.0000,0.0000,1.0,1.0000
Pclass,891.0,2.308642,0.836071,1.00,2.0000,3.0000,3.0,3.0000
Age,714.0,29.699118,14.526497,0.42,20.1250,28.0000,38.0,80.0000
SibSp,891.0,0.523008,1.102743,0.00,0.0000,0.0000,1.0,8.0000
Parch,891.0,0.381594,0.806057,0.00,0.0000,0.0000,0.0,6.0000
Fare,891.0,32.204208,49.693429,0.00,7.9104,14.4542,31.0,512.3292


<p style="font-weight: bold; font-style: italic;">Description of numerical columns in train data</p>


In [146]:
ana_train_data.describe().T

,count,mean,std,min,25%,50%,75%,max
Age,714.0,29.699118,14.526497,0.42,20.1250,28.0000,38.0,80.0000
Fare,891.0,32.204208,49.693429,0.00,7.9104,14.4542,31.0,512.3292


<p style="font-weight: bold; font-style: italic;">Description of categorical columns in train data</p>


In [147]:
ana_train_data.describe(exclude=float).T

,count,unique,top,freq
Survived,891,2,0,549
Pclass,891,3,3,491
Name,891,891,"Braund, Mr. Owen Harris",1
Sex,891,2,male,577
SibSp,891,7,0,608
Parch,891,7,0,678
Ticket,891,681,347082,7
Cabin,204,147,B96 B98,4
Embarked,889,3,S,644


## 5) **Data split**

<p>Separate target from the data and then split the data into train and validation</p>


In [148]:
# Target name.
TARGET_NAME = "Survived"

# Separating target column from the feature columns.
X = train_df.copy()
y = X.pop(TARGET_NAME).values.reshape(-1, 1)

# View the shape X and y.
print(f"""
X shape: {X.shape}
y shape: {y.shape}
""")


X shape: (891, 10)
y shape: (891, 1)



In [149]:
# 25 percent will be for test data.
test_size = 0.25

# 🪓 Splitting data into train and validation data.
train_x, valid_x, train_y, valid_y = train_test_split(X, y, test_size=test_size)

# View the shape of the data.
print(f"""
train_x shape: {train_x.shape}
train_y shape: {train_y.shape}
valid_x shape: {valid_x.shape}
valid_y shape: {valid_y.shape}
""")


train_x shape: (668, 10)
train_y shape: (668, 1)
valid_x shape: (223, 10)
valid_y shape: (223, 1)



## 6) **Data cleaning**

<p>Modifying or removing data that is inaccurate, duplicate, incomplete, incorrectly formatted, or corrupted</p>


### i) Check the missing values (NaN)

<p>Search for missing value in the dataframes</p>


In [150]:
def check_n_nan(data: pd.DataFrame, only_nan: bool = True) -> pd.DataFrame:
    """
    Represents the summary of missing data in the
    data frame.

    Parameters:
    ----------
    data: A data frame to check for missing data.

    return: A dataframe with number of missing value (NaN) and the percentage.
    """
    # Compute the total NaN for each column.
    total_nan = data.isna().sum()

    if only_nan:
        total_nan = total_nan[total_nan > 0]

    # Columns
    column_names = total_nan.index

    # Compute the percentage of missing data.
    percentage = (
        ((total_nan / data.shape[0]) * 100).round(2).astype(str).map(lambda x: x + "%")
    )

    return pd.DataFrame(
        {"columns": column_names, "total_nan": total_nan, "percent_nan": percentage}
    ).reset_index(drop=True)

In [151]:
def plot_nan(
    data: pd.DataFrame,
    name: str | None = None,
    only_nan: bool = True,
    use_percent=False,
) -> None:
    nan_df = check_n_nan(data, only_nan)
    nan_df["percent_nan"] = (
        nan_df["percent_nan"].str.replace("%", "").astype(float).round().astype(int)
    )

    radius = 5
    y_name = "percent_nan" if use_percent else "total_nan"
    startup_title = "Percentage" if use_percent else "Number"
    y_title = f"{startup_title} of missing values"
    y_name = alt.Y(y_name).title(y_title)
    name = name if name else "a dataset"
    title = f"{startup_title} Of Missing values In {name.title()}"
    subtitle = f"A chart represents missing data in {name}"

    base = (
        alt.Chart(
            nan_df,
            title=alt.Title(
                title,
                subtitle=subtitle,
                anchor="start",
            ),
        )
        .mark_bar(cornerRadiusTopLeft=radius, cornerRadiusTopRight=radius)
        .encode(
            x=alt.X("columns").title("Columns names"),
            y=y_name,
            color=alt.Color("columns")
            .scale(scheme="lightgreyred", reverse=True)
            .legend(orient="right"),
            tooltip=["columns", "total_nan", "percent_nan"],
            text="columns",
        )
        .configure_axis(grid=False)
        .configure_view(stroke=None)
    )

    return base

#### a) Observing missing data in train and validation data


##### One) X train dataset


In [152]:
check_n_nan(train_x)

,columns,total_nan,percent_nan
0,Age,135,20.21%
1,Cabin,524,78.44%
2,Embarked,2,0.3%


In [153]:
plot_nan(train_x, name="x train dataset", only_nan=False)

<VegaLite 5 object>

If you see this message, it means the renderer has not been properly enabled
for the frontend that you are using. For more information, see
https://altair-viz.github.io/user_guide/display_frontends.html#troubleshooting


**conclusion:** X train data sample contains missing data of 21.11%(141) of Age column, 78.29%(523) of cabin column and 0.3%(2) of Embarked column are the amount of missing data in the dataset from the data size of 668.


##### Two) X validation dataset


In [154]:
check_n_nan(valid_x)

,columns,total_nan,percent_nan
0,Age,42,18.83%
1,Cabin,163,73.09%


In [155]:
plot_nan(valid_x, name="x validation dataset", only_nan=False, use_percent=True)

<VegaLite 5 object>

If you see this message, it means the renderer has not been properly enabled
for the frontend that you are using. For more information, see
https://altair-viz.github.io/user_guide/display_frontends.html#troubleshooting


**conclusion:** In X validation data sample there are 16.14%(36) of Age column and 73.54%(164) of cabin column are the amount of missing data in the dataset from the data size of 223.


#### b) Observing missing data in test dataset


In [156]:
check_n_nan(test_df)

,columns,total_nan,percent_nan
0,Age,86,20.57%
1,Fare,1,0.24%
2,Cabin,327,78.23%


In [157]:
plot_nan(test_df, name="test dataset", only_nan=False, use_percent=False)

<VegaLite 5 object>

If you see this message, it means the renderer has not been properly enabled
for the frontend that you are using. For more information, see
https://altair-viz.github.io/user_guide/display_frontends.html#troubleshooting


**conclusion:** The test data frame contains missing data of 20.57%(86) of Age column, 78.23%(327) of cabin column and 0.24%(1) of Fare column are the amount of missing data in the dataset from the data size of 418.


### ii) Handling the missing values (NaN)

<p>Filling missing value in the dataframes</p>


In [158]:
class FillNaN:
    def __init__(self, dfs: List[pd.DataFrame], fill_with: Dict[str, Any]) -> None:
        self.dfs = dfs
        self.fill_with = fill_with

    @property
    def fill(self) -> List[pd.DataFrame]:
        """
        Fills missing values in the data frames.
        :return: A tuple of data frames
        """
        lst = []
        for df in self.dfs:
            for col, strat in self.fill_with.items():
                df = self.strategy(df, strat=strat, col_name=col)
            lst.append(df)
        return lst

    @staticmethod
    def strategy(df: pd.DataFrame, strat: str, col_name: str) -> pd.DataFrame:
        match strat:
            case "mean":
                df[col_name] = df[col_name].fillna(df[col_name].mean())
            case "mode":
                df[col_name] = df[col_name].fillna(df[col_name].mode()[0])
            case "min":
                df[col_name] = df[col_name].fillna(df[col_name].min())
            case "max":
                df[col_name] = df[col_name].fillna(df[col_name].max())
            case "median":
                df[col_name] = df[col_name].fillna(df[col_name].median())
            case "bfill":
                df[col_name] = df[col_name].bfill()
            case "ffill":
                df[col_name] = df[col_name].ffill()
            case "drop":
                df = df.drop(col_name, axis=1)
            case _:
                raise ValueError(
                    f"Ooh, strategy has received invalid value `{strat}` for strat"
                )
        return df

Filling the missing value in age with mean, Fare with ffill, Embarked with mode and Cabin with drop.


In [159]:
fill = FillNaN(
    [train_x, valid_x, test_df],
    {"Age": "mean", "Fare": "ffill", "Embarked": "mode", "Cabin": "drop"},
)

In [160]:
clean_train_x, clean_valid_x, clean_test_df = fill.fill

Check if there still any missing values in the data.


**Clean train x**


In [161]:
check_n_nan(clean_train_x, only_nan=False)

,columns,total_nan,percent_nan
0,Pclass,0,0.0%
1,Name,0,0.0%
2,Sex,0,0.0%
3,Age,0,0.0%
4,SibSp,0,0.0%
5,Parch,0,0.0%
6,Ticket,0,0.0%
7,Fare,0,0.0%
8,Embarked,0,0.0%


**Clean valid x**


In [162]:
check_n_nan(clean_valid_x, only_nan=False)

,columns,total_nan,percent_nan
0,Pclass,0,0.0%
1,Name,0,0.0%
2,Sex,0,0.0%
3,Age,0,0.0%
4,SibSp,0,0.0%
5,Parch,0,0.0%
6,Ticket,0,0.0%
7,Fare,0,0.0%
8,Embarked,0,0.0%


**Clean test df**


In [163]:
check_n_nan(clean_test_df, only_nan=False)

,columns,total_nan,percent_nan
0,Pclass,0,0.0%
1,Name,0,0.0%
2,Sex,0,0.0%
3,Age,0,0.0%
4,SibSp,0,0.0%
5,Parch,0,0.0%
6,Ticket,0,0.0%
7,Fare,0,0.0%
8,Embarked,0,0.0%


### iii) Handling the categorical columns

<p>Encode categorical column to integer and drop high cardinality columns in the dataframes</p>


In [164]:
class CategoryEncoder:
    def __init__(
        self,
        datasets: List[pd.DataFrame],
        handle_with: Dict[str, str],
    ) -> None:
        self.datasets = datasets
        self.handle_with = handle_with

    def strategy(self, df: pd.DataFrame, col: str, strat: str) -> pd.DataFrame:
        match strat:
            case "label_encode":
                series = df[col].astype("category").cat.codes
                df[col] = series
            case "one_hot_encode":
                data = pd.get_dummies(df[col], columns=list(df[col].unique())).astype(
                    int
                )
                df = pd.concat([df, data], axis=1)
                df = df.drop(col, axis=1)
            case "drop":
                df = df.drop(col, axis=1)
            case _:
                raise ValueError()

        return df

    def encode(self) -> List[pd.DataFrame]:
        """
        Encode categorical columns to integer.
        :return: A tuple of data frames
        """

        datasets = []

        for dataset in self.datasets:
            for col, strat in self.handle_with.items():
                dataset = self.strategy(dataset, col, strat)
            datasets.append(dataset)

        return datasets

In [165]:
# First five rows from the clean train x.
clean_train_x.head()

,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
PassengerId,,,,,,,,,
270,1,"Bissette, Miss. Amelia",female,35.000000,0,0,PC 17760,135.6333,S
532,3,"Toufik, Mr. Nakli",male,30.081144,0,0,2641,7.2292,C
131,3,"Drazenoic, Mr. Jozef",male,33.000000,0,0,349241,7.8958,C
21,2,"Fynney, Mr. Joseph J",male,35.000000,0,0,239865,26.0000,S
864,3,"Sage, Miss. Dorothy Edith ""Dolly""",female,30.081144,8,2,CA. 2343,69.5500,S


In [166]:
# Exclude the int and float from the data.
cat_df = clean_train_x.select_dtypes(exclude=[int, float])
cat_df.head()

,Name,Sex,Ticket,Embarked
PassengerId,,,,
270,"Bissette, Miss. Amelia",female,PC 17760,S
532,"Toufik, Mr. Nakli",male,2641,C
131,"Drazenoic, Mr. Jozef",male,349241,C
21,"Fynney, Mr. Joseph J",male,239865,S
864,"Sage, Miss. Dorothy Edith ""Dolly""",female,CA. 2343,S


In [167]:
# Fetch the categorical columns.
cat_columns = list(cat_df.columns)
cat_columns

['Name', 'Sex', 'Ticket', 'Embarked']

Check for columns with high cardinality column in the data.


In [168]:
clean_train_x[cat_columns].nunique()

Name        668
Sex           2
Ticket      543
Embarked      3
dtype: int64

In [169]:
# Instantiate the CategoryEncoder
encoder = CategoryEncoder(
    [clean_train_x, clean_valid_x, clean_test_df],
    {
        "Sex": "label_encode",
        "Embarked": "one_hot_encode",
        "Name": "drop",
        "Ticket": "drop",
    },
)

# Encode the column
num_train_x, num_valid_x, num_test_df = encoder.encode()

In [170]:
num_train_x.dtypes

Pclass      int64
Sex          int8
Age       float64
SibSp       int64
Parch       int64
Fare      float64
C           int64
Q           int64
S           int64
dtype: object

## 7) **Data visualisation**

<p>Represeting the data in graphical way like making chart</p>


In [171]:
visual_df = pd.concat([clean_train_x, clean_valid_x])
visual_df["Survived"] = np.concatenate([train_y, valid_y])
visual_df.head()

,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Survived
PassengerId,,,,,,,,,,
270,1,"Bissette, Miss. Amelia",0,35.000000,0,0,PC 17760,135.6333,S,1
532,3,"Toufik, Mr. Nakli",1,30.081144,0,0,2641,7.2292,C,0
131,3,"Drazenoic, Mr. Jozef",1,33.000000,0,0,349241,7.8958,C,0
21,2,"Fynney, Mr. Joseph J",1,35.000000,0,0,239865,26.0000,S,0
864,3,"Sage, Miss. Dorothy Edith ""Dolly""",0,30.081144,8,2,CA. 2343,69.5500,S,0


In [172]:
alt.renderers.enable("html")


class Graph:
    def __init__(
        self,
        data: pd.DataFrame,
        map_values: Optional[Dict[str, Dict[Any, Any]]] = None,
        theme: str = "light",
    ) -> None:
        data = data.copy()

        if map_values:
            for key, value in map_values.items():
                data[key] = data[key].map(lambda col: value[col])

        self.__data = data
        self.__theme = theme

    def setting_colors(
        self, x: str, color: List[str] | str
    ) -> Tuple[pd.DataFrame, alt.Color]:

        # Color map.
        if isinstance(color, list):
            alt_color = (
                alt.Color(x).scale(range=color, reverse=True).legend(orient="right")
            )
        else:
            alt_color = (
                alt.Color(x).scale(scheme=color, reverse=True).legend(orient="right")
            )
        return alt_color

    def normalized_bar(self, x, y, hue, color, strategy="mean"):
        data = self.__data.copy()

        d = data.groupby([y, hue])[x].agg(strategy).reset_index()

        alt_color = self.setting_colors(hue, color=color)

        bars = (
            alt.Chart(d)
            .mark_bar()
            .encode(
                x=alt.X(f"{strategy}({x}):Q").stack("normalize"),
                y=alt.Y(f"{y}:N"),
                color=alt_color,
            )
        )

        text = (
            alt.Chart(d)
            .mark_text(dx=-15, dy=3, color="white")
            .encode(
                x=alt.X(f"{strategy}({x}):Q").stack("normalize"),
                y=alt.Y(f"{y}:N"),
                detail=f"{hue}:N",
                text=alt.Text(f"{strategy}({x}):Q", format=".1f"),
            )
        )

        return bars + text

    def col_bar_char(
        self,
        x_column: str,
        group_by: str,
        color: List | str = "paired",
        width: int = 200,
        height: int = 200,
        title: Optional[str] = None,
        subtitle: Optional[str] = None,
        x_title: str | None = None,
        y_title: str | None = None,
    ):
        alt_color = self.setting_colors(x_column, color=color)

        data = self.__data

        x_title = x_title.title() if x_title else x_column.title()
        y_title = y_title.title() if y_title else "Number of Passengers"

        source = data.groupby(group_by)[x_column].value_counts().reset_index()
        source["percent"] = (source["count"] / visual_df.shape[0]).round(4) * 100

        grouped_bars = (
            alt.Chart(source)
            .mark_bar(cornerRadius=10)
            .encode(
                x=alt.X(x_column, title=x_title),
                y=alt.Y(
                    "count:Q",
                    title=y_title,
                    axis=alt.Axis(
                        grid=True, gridDash=[1, 2]  # Ensure grid lines are visible
                    ),
                ),
                color=alt_color,
                column=group_by,
            )
            .encode(
                tooltip=[
                    alt.Tooltip(x_column, title=x_title),
                    alt.Tooltip("count", title=y_title),
                    alt.Tooltip("percent:Q", format=".2f", title="Percentage (%)"),
                ],
            )
            .properties(
                width=width,
                height=height,
                title=alt.Title(
                    title if title else "",
                    fontSize=18,
                    subtitle=subtitle if subtitle else "",
                    subtitleFontSize=12,
                ),
            )
        )

        chart = grouped_bars
        return chart

    def bar_chart(
        self,
        x_column: str,
        color: List | str = "paired",
        width: int = 200,
        height: int = 200,
        title: Optional[str] = None,
        subtitle: Optional[str] = None,
        x_title: str | None = None,
        y_title: str | None = None,
    ):

        alt_color = self.setting_colors(x_column, color=color)

        data = self.__data.copy()

        x_title = x_title.title() if x_title else x_column.title()
        y_title = y_title.title() if y_title else "Number of Passengers"

        # Return the alt chart
        bar_layer = (
            alt.Chart(data)
            .mark_bar(cornerRadius=10)
            .encode(
                x=alt.X(
                    x_column,
                    title=x_title,
                    axis=alt.Axis(grid=False, labelFontWeight="bold"),
                ),
                y=alt.Y(
                    "count()",
                    title=y_title,
                    axis=alt.Axis(grid=False, labelFontWeight="bold"),
                ),
                color=alt_color.legend(orient="right"),
                tooltip=[
                    alt.Tooltip(x_column, title=x_title),
                    alt.Tooltip("count()", title=y_title),
                ],
            )
            .properties(
                width=width,
                height=height,
                title=alt.Title(
                    title if title else "",
                    fontSize=18,
                    subtitle=subtitle if subtitle else "",
                    subtitleFontSize=12,
                    # dy=-10,
                ),
            )
        )

        text_layer = (
            alt.Chart(data)
            .mark_text(
                align="center",
                baseline="bottom",  # Align text to the bottom of its bounding box
                dy=-5,  # Adjust vertical position (negative for above the bar)
                fontSize=12,
                fontWeight="bold",
                color="white",
            )
            .encode(x=x_column, y="count()", text=alt.Text("count()"))
        )

        charts = bar_layer + text_layer

        return charts

    def donut_chart(
        self,
        x: str,
        inner_radius: int = 50,
        color: List | str = "paired",
        width: int = 200,
        height: int = 200,
    ):

        data = self.__data

        value_counts = data[x].value_counts().reset_index()
        value_counts["count"] = ((value_counts["count"] / data.shape[0]) * 100).round(2)

        donut_chart = (
            alt.Chart(value_counts)
            .mark_arc(innerRadius=inner_radius)
            .encode(
                theta="count",
                color=f"{x}:N",
                tooltip=[
                    alt.Tooltip(f"{x}".title()),
                    alt.Tooltip("count:Q", format=".2f", title="Percentage (%)"),
                ],
            )
            .properties(
                width=width,
                height=height,
            )
        )

        return donut_chart

    def scatter_chart(
        self,
        x: str,
        y: str,
        color_with: Optional[str] = None,
        width: int = 300,
        height: int = 300,
        title: Optional[str] = None,
        subtitle: Optional[str] = None,
        color: List[str] | str = "dark2",
        use_subplot: bool = False,
    ):
        alt_color = None

        if color_with:
            alt_color = self.setting_colors(color_with, color=color)

        tooltips = (
            [alt.Tooltip(x), alt.Tooltip(y), alt.Tooltip(color_with)]
            if color_with
            else [alt.Tooltip(x), alt.Tooltip(y)]
        )

        encode_param = (
            dict(x="%s:Q" % x, y="%s:Q" % y, color=alt_color, tooltip=tooltips)
            if color_with
            else dict(x="%s:Q" % x, y="%s:Q" % y, tooltip=tooltips)
        )

        data = self.__data

        # Scatter Plot
        points = (
            alt.Chart(data)
            .mark_point()
            .encode(**encode_param)
            .properties(
                width=width,
                height=height,
                title=alt.Title(
                    title if title else "",
                    fontSize=18,
                    subtitle=subtitle if subtitle else "",
                    subtitleFontSize=12,
                    dx=10, dy=-10
                ),
            )
        )

        if use_subplot:
            return points

        points = (
            points.configure_legend(
                strokeColor="gray",
                fillColor="#EEEEEE",
                padding=10,
                cornerRadius=10,
                orient="top-right",
            )
            .configure_axis(gridDash=[5, 2, 5], gridColor="lightgrey")
            .configure_view(stroke=None)
        )

        return points.display(theme=self.__theme)

    def _get_color_config(self, color: str | None):
        data = self.__data
        is_col = color in data.columns
        return {
            "condition": color if is_col else alt.Undefined,
            "value": color if (color and not is_col) else alt.Undefined
        }

    def hist_chart(
        self,
        x: str,
        width=200,
        height=200,
        n_bins=10,
        color: str | None = None,
        mean_color: str = "red",
        median_color: str = "green"
    ):
        data = self.__data
        color_map = self._get_color_config(color)

        # 1. Create the main histogram
        chart = (
            alt.Chart(data)
            .mark_bar(color=color_map['value'])
            .encode(
                x=alt.X(f"{x}:Q", bin=alt.Bin(maxbins=n_bins), title=x),
                y=alt.Y("count()", title="Frequency"),
                tooltip=[x, "count()"],
                color=color_map["condition"],
            )
            .properties(width=width, height=height)
        )

        # 2. Create a long-format DataFrame for rules
        rules_df = pd.DataFrame({
            'Value': [data[x].mean(), data[x].median()],
            'Metric': ['Mean', 'Median']
        })

        # 2. Smooth Density Curve Layer
        curve = (
            alt.Chart(data)
            .transform_density(
                density=x,
                as_=[x, 'density']
            )
            .mark_line(color='#010', size=2)  # Adjust line appearance here
            .encode(
                x=f"{x}:Q",
                y='density:Q',
                tooltip=[x, "count()"]
            )
        )

        # 3. Create a single rule layer with a legend map
        rules = (
            alt.Chart(rules_df)
            .mark_rule(size=2)
            .encode(
                x='Value:Q',
                color=alt.Color(
                    'Metric:N',
                    scale=alt.Scale(domain=['Mean', 'Median'], range=[mean_color, median_color]),
                    title='Statistics'
                ),
                tooltip=[
                    alt.Tooltip('Value', title='Metric'),
                ]
            )
        )

        # 4. Combine layers
        return alt.layer(chart, curve, rules).resolve_scale(y='independent')



    def box_chart(
        self, x: str, width: int = 200, height: int = 50,
        color: str | None = None,
    ):
        data = self.__data
        color_map = self._get_color_config(color)

        box_chart = (
            alt.Chart(data)
            .mark_boxplot(color=color_map['value'])
            .encode(
                alt.X("%s:Q" % x).scale(zero=False),
                color=color_map['condition'],
            )
            .properties(width=width)
        )

        return box_chart

    def categorical_grouped_chart(
        self,
        x: str,
        hue: str,
        colors: List[str],
        target_colors: List[str],
        title: str,
        subtitle: str,
    ):

        x_bar = self.bar_chart(x, color=colors)
        x_donut = self.donut_chart(x)
        x_vs_hue = self.col_bar_char(hue, group_by=x, color=target_colors)

        return (
            ((x_bar | x_donut) & x_vs_hue)
            .resolve_scale(color="independent")
            .configure_view(stroke=None)
            .properties(
                title=alt.Title(title, subtitle=subtitle, fontSize=15, dx=10, dy=-10)
            )
            .display(theme=self.__theme)
        )

    def categorical_charts(
        self,
        x: str,
        colors: List[str],
        title: str,
        subtitle: str,
    ):

        x_bar = self.bar_chart(x, color=colors)
        x_donut = self.donut_chart(x)

        return (
            (x_bar | x_donut)
            .configure_view(stroke=None)
            .properties(
                title=alt.Title(title, subtitle=subtitle, fontSize=15, dx=10, dy=-10)
            )
            .display(theme=self.__theme)
        )

    def continuous_charts(
        self,
        x: str,
        title: str,
        subtitle: str,
        color: str = None,
        width: int = 500,
        hist_height: int = 200,
        box_height: int = 50,
        n_bins: int = 10
    ):

        x_bar = self.hist_chart(
            x, color=color, 
            width=width, height=hist_height,
            n_bins=n_bins
        )
        x_donut = self.box_chart(
            x, 
            width=width, 
            height=box_height,
            color=color
        )

        return (
            (x_bar & x_donut)
            .configure_view(stroke=None)
            .properties(
                title=alt.Title(title, subtitle=subtitle, fontSize=15, dx=10, dy=-10)
            )
            .display(theme=self.__theme)
        )

In [173]:
graph = Graph(
    visual_df,
    map_values={
        "Sex": {0: "Female", 1: "Male"},
        "Survived": {0: "Didn't survive", 1: "survived"},
        "Embarked": {"C": "Cherbourg", "Q": "Queenstown", "S": "Southampton"},
        "Pclass": {1: "First class", 2: "Second class", 3: "Third class"},
    },
    theme="light",
)

**Target (Survived) Visualization**

In [174]:
SURVIVED_COLORS = ["green", "red"]
title = "Passenger Survivor Visualization"
SUBTITLE = "Source: Titanic crash"

In [175]:
graph.categorical_charts(
    x=TARGET_NAME, colors=SURVIVED_COLORS, subtitle=SUBTITLE, title=title
)

alt.HConcatChart(...)

**Sex Visualization**

In [176]:
SEX_COLORS = ["blue", "yellow"]
title = "Passengers Sex And Survivor With Sex"
SEX = "Sex"

In [177]:
graph.categorical_grouped_chart(
    SEX,
    TARGET_NAME,
    colors=SEX_COLORS,
    target_colors=SURVIVED_COLORS,
    title=title,
    subtitle=SUBTITLE,
)

alt.VConcatChart(...)

**Summary analysis**: A high number of men were on the ship but few of them were survived with percentage of 12.23% and 26.15% for women survived and 52.53% of men died and 9.09 of women died.

**Embarked Visualization**

In [178]:
EMBARKED_COLORS = ["#6494AA", "#C69DD2", "#EA1744"]
title = "Passengers Embarkment And Survivor By Embarkment"
EMBARKED = "Embarked"

In [179]:
graph.categorical_grouped_chart(
    EMBARKED,
    TARGET_NAME,
    colors=EMBARKED_COLORS,
    target_colors=SURVIVED_COLORS,
    title=title,
    subtitle=SUBTITLE,
)

alt.VConcatChart(...)

**Summary analysis:** Embarkment of the ship from Cherbourg, Queenstown and Southampton in the analysis 
the data shows that a high number of passenger were from southampton and about 47.92% in southampton
did not survive the crash of the ship but 24.58% survived and Queenstowns with low percentage of 8.64 and 3.37% survived and 5.27 didn't survive and Cherbourg with 18.86 passengers with 10.44% survivors and 8.42% who did not 
survive the crush

**Pclass Visualsation**

In [180]:
PCLASS_COLORS = ["#6414AD", "#C65DA2", "#AAF704"]
title = "Passengers class And Survivor By Their Classes"
PCLASS = "Pclass"

In [181]:
graph.categorical_grouped_chart(
    PCLASS,
    TARGET_NAME,
    colors=PCLASS_COLORS,
    target_colors=SURVIVED_COLORS,
    title=title,
    subtitle=SUBTITLE,
)

alt.VConcatChart(...)

**Summary analysis:** Passenger class, A high number of passenger were in third class with 55.11% and a low percentage of 20.65% in second class and first class with 24.24%, A high percentage of survivors in first class with 15.26% with  and a high percentage of those who didn't survive in third class with 41.75%

**Age visualization**

In [182]:
graph.continuous_charts(
    "Age", 
    "Distribution Of Passenger Age", 
    "Spread of passenger age", 
    n_bins=50,
    hist_height=400
    )

alt.VConcatChart(...)

**Fare visualization**

In [183]:
graph.continuous_charts(
    "Fare", 
    "Distribution Of Passenger Fare", 
    "Spread of passenger fare", 
    n_bins=50,
    hist_height=400
    )

alt.VConcatChart(...)

**Relative Of Age With Fare**

In [184]:
graph.scatter_chart(
    x="Age",
    y="Fare",
    color_with="Survived",
    color=["green", "red"],
    width=500,
    title="Age Vs Fare",
    subtitle="Survival rate based on age and fare"
)

alt.Chart(...)

## 8) **Model construction**

Creating survival predictive model\
<a href="#top">scroll to up &gt;</a>

In [185]:
RANDOM_SEED = 42

**Explaining the metric**\
**Accuracy:** The proportion of correctly classified instances among the total instances.\
**Precision:** The proportion of true positive predictions among all positive predictions.\
**Recall:** The proportion of true positive predictions among all actual positive instances.\
**F1 Score:** The harmonic mean of precision and recall.\
**Validation Accuracy:** The accuracy of the model on the validation set.\
**Validation Precision:** The precision of the model on the validation set.\
**Validation Recall:** The recall of the model on the validation set.\
**Validation F1 Score:** The F1 score of the model on the validation set.



In [186]:
def model_measurements(
    model: BaseEstimator | ClassifierMixin,
    x: np.ndarray,
    y: np.ndarray,
    valid_x: np.ndarray | None = None,
    valid_y: np.ndarray | None = None,
) -> pd.DataFrame:
    """
    Computes the model measurements.

    Args
    ----
    model: A trained model.
    x: The input features.
    y: The true values.
    valid_x: The validation input features.
    valid_y: The validation true values.

    :return: A dataframe with model measurements.
    """

    # Fit the model on the training data.
    model.fit(x, y.ravel())

    predictions = model.predict(x)
    valid_predictions = model.predict(valid_x) if valid_x is not None else None

    # Compute the accuracy, precision, recall and f1 score.
    accuracy = accuracy_score(y, predictions)
    precision = precision_score(y, predictions)
    recall = recall_score(y, predictions)
    f1 = f1_score(y, predictions)

    measurements = {
        "Model": model.__class__.__name__,
        "Accuracy": accuracy,
        "Precision": precision,
        "Recall": recall,
        "F1 Score": f1,
    }

    measurements_df = pd.DataFrame([measurements])

    if valid_predictions is not None:
        valid_accuracy = accuracy_score(valid_y, valid_predictions)
        valid_precision = precision_score(valid_y, valid_predictions)
        valid_recall = recall_score(valid_y, valid_predictions)
        valid_f1 = f1_score(valid_y, valid_predictions)

        
        measurements_df["Validation Accuracy"] = valid_accuracy
        measurements_df["Validation Precision"] = valid_precision
        measurements_df["Validation Recall"] = valid_recall
        measurements_df["Validation F1 Score"] = valid_f1
        

    return measurements_df

In [187]:
def model_confusion_matrix(
    model: BaseEstimator | ClassifierMixin,
    x: np.ndarray,
    y: np.ndarray,
    valid_x: np.ndarray | None = None,
    valid_y: np.ndarray | None = None,
) -> Dict[str, np.ndarray]:
    """
    Computes the confusion matrix for the model.

    Args
    ----
    model: A trained model.
    x: The input features.
    y: The true values.
    valid_x: The validation input features.
    valid_y: The validation true values.

    :return: A dictionary with confusion matrices for training and validation data.
    """
    model.fit(x, y.ravel())
    
    predictions = model.predict(x)
    valid_predictions = model.predict(valid_x) if valid_x is not None else None

    confusion_matrices = {
        "Training Confusion Matrix": confusion_matrix(y, predictions)
    }

    if valid_predictions is not None:
        confusion_matrices["Validation Confusion Matrix"] = confusion_matrix(
            valid_y, valid_predictions
        )

    return confusion_matrices

In [188]:
def measure_models(
    models: List[BaseEstimator | ClassifierMixin],
    x: np.ndarray,
    y: np.ndarray,
    valid_x: np.ndarray | None = None,
    valid_y: np.ndarray | None = None,
    sort_by: Literal[
        "Accuracy",
        "Precision",
        "Recall",
        "F1 Score",
        "Validation Accuracy",
        "Validation Precision",
        "Validation Recall",
        "Validation F1 Score",
    ] = "Validation Accuracy"
) -> pd.DataFrame:
    """
    Measures the performance of multiple models.

    Args
    ----
    models: A list of trained models.
    x: The input features.
    y: The true values.
    valid_x: The validation input features.
    valid_y: The validation true values.

    :return: A dataframe with model measurements for all models.
    """
    measurements_list = []

    for model in models:
        measurements_df = model_measurements(model, x, y, valid_x, valid_y)
        measurements_list.append(measurements_df)

    return (
        pd.concat(measurements_list, ignore_index=True)
        .sort_values(by=sort_by, ascending=False).reset_index(drop=True)
    )

In [202]:
def plot_models_performance(
    measurements_df: pd.DataFrame,
    metric: Literal[
        "Accuracy",
        "Precision",
        "Recall",
        "F1 Score",
        "Validation Accuracy",
        "Validation Precision",
        "Validation Recall",
        "Validation F1 Score",
    ],
) -> alt.Chart:
    """
    Plots the performance of models based on a specified metric.

    Args
    ----
    measurements_df: A dataframe containing model measurements.
    metric: The metric to plot.

    :return: An Altair chart visualizing the model performance.
    """
    
    px.bar(
        measurements_df,
        x="Model",
        y=metric,
        color="Model",
        title=f"Model Performance Based on {metric}",
        labels={"Model": "Model", metric: metric},
        text_auto=True,
    ).update_layout(
        xaxis_title="Model",
        yaxis_title=metric,
        legend_title="Model",
        template="plotly_white",
    ).show()


In [210]:
def feature_importance(
    model: BaseEstimator | ClassifierMixin,
    x: pd.DataFrame,
    y: np.ndarray,
    feature_names: str,
    valid: pd.DataFrame | None = None,
    valid_y: np.ndarray | None = None,
    metric: Literal[
        "Accuracy",
        "Precision",
        "Recall",
        "F1 Score",
        "Validation Accuracy",
        "Validation Precision",
        "Validation Recall",
        "Validation F1 Score",
    ] = "Validation Accuracy"
) -> pd.DataFrame:
    """
    Computes the feature importance for a given model.

    Args
    ----
    model: A trained model with feature importance attribute.
    x: The input features.
    y: The target values.
    valid: The validation features.
    valid_y: The validation target values.

    :return: A dataframe containing the top_n features and their importance scores.
    """
    x = x.copy().loc[:, feature_names].values.reshape(-1, 1)
    valid = valid.copy().loc[:, feature_names].values.reshape(-1, 1) if valid is not None else None

    model.fit(x, y.ravel())

    if valid is not None and valid_y is not None:
        valid_predictions = model.predict(valid)
        match metric:
            case "Validation Accuracy":
                result = accuracy_score(valid_y, valid_predictions)
            case "Validation Precision":
                result = precision_score(valid_y, valid_predictions)
            case "Validation Recall":
                result = recall_score(valid_y, valid_predictions)
            case "Validation F1 Score":
                result = f1_score(valid_y, valid_predictions)
            case _:
                result = accuracy_score(valid_y, valid_predictions)
    else:
        predictions = model.predict(x)
        match metric:
            case "Accuracy":
                result = accuracy_score(y, predictions)
            case "Precision":
                result = precision_score(y, predictions)
            case "Recall":
                result = recall_score(y, predictions)
            case "F1 Score":
                result = f1_score(y, predictions)
            case _:
                result = accuracy_score(y, predictions)

    return {
        feature_names: result
    }


def features_importance_summary(
    model: BaseEstimator | ClassifierMixin,
    x: pd.DataFrame,
    y: np.ndarray,
    valid_x: pd.DataFrame | None = None,
    valid_y: np.ndarray | None = None,
    metric: Literal[
        "Accuracy",
        "Precision",
        "Recall",
        "F1 Score",
        "Validation Accuracy",
        "Validation Precision",
        "Validation Recall",
        "Validation F1 Score",
    ] = "Validation Accuracy"
) -> pd.DataFrame:
    """
    Computes the feature importance summary for a given model.

    Args
    ----
    model: A trained model with feature importance attribute.
    x: The input features.
    y: The target values.
    valid_x: The validation features.
    valid_y: The validation target values.

    :return: A dataframe containing the feature importance summary.
    """
    feature_names = x.columns.tolist()
    importance_scores = []

    for feature in feature_names:
        score = feature_importance(model, x, y, feature, valid_x, valid_y, metric)
        importance_scores.append((feature, list(score.values())[0]))

    importance_df = pd.DataFrame(importance_scores, columns=["Feature", "Importance"])
    
    return importance_df.sort_values(by="Importance", ascending=False).reset_index(drop=True)

def plot_feature_importance(
    importance_df: pd.DataFrame,
    metric: Literal[
        "Accuracy",
        "Precision",
        "Recall",
        "F1 Score",
        "Validation Accuracy",
        "Validation Precision",
        "Validation Recall",
        "Validation F1 Score",
    ] = "Validation Accuracy"
) -> alt.Chart:
    """
    Plots the feature importance based on a specified metric.

    Args
    ----
    importance_df: A dataframe containing feature importance scores.
    metric: The metric to plot.

    :return: An Altair chart visualizing the feature importance.
    """
    
    px.bar(
        importance_df,
        x="Feature",
        y="Importance",
        color="Feature",
        title=f"Feature Importance Based on {metric}",
        labels={"Feature": "Feature", "Importance": metric},
        text_auto=True,
    ).update_layout(
        xaxis_title="Feature",
        yaxis_title=metric,
        legend_title="Feature",
        template="plotly_white",
    ).show()

### i) **Model performance measurement**

In [217]:
all_models = [
    tree.DecisionTreeClassifier(random_state=RANDOM_SEED),
    tree.ExtraTreeClassifier(random_state=RANDOM_SEED),
    ensemble.RandomForestClassifier(random_state=RANDOM_SEED),
    ensemble.AdaBoostClassifier(ensemble.RandomForestClassifier(random_state=RANDOM_SEED)),
    ensemble.GradientBoostingClassifier(random_state=RANDOM_SEED),
    ensemble.BaggingClassifier(random_state=RANDOM_SEED),
    ensemble.ExtraTreesClassifier(random_state=RANDOM_SEED),
    ensemble.HistGradientBoostingClassifier(random_state=RANDOM_SEED),
    ensemble.VotingClassifier(estimators=[
        ("rf", ensemble.RandomForestClassifier(random_state=RANDOM_SEED)),
        ("gb", ensemble.GradientBoostingClassifier(random_state=RANDOM_SEED)),
        ("dt", tree.DecisionTreeClassifier(random_state=RANDOM_SEED)),
        ("gaussian_process", gaussian_process.GaussianProcessClassifier(random_state=RANDOM_SEED)),
        ("kn", neighbors.KNeighborsClassifier()),
        ("svm", svm.SVC(random_state=RANDOM_SEED)),
        ("mlp", neural_network.MLPClassifier(random_state=RANDOM_SEED)),
        ("logistic", linear_model.LogisticRegression(max_iter=150, random_state=RANDOM_SEED)),
        ("naive_bayes", naive_bayes.GaussianNB()),
    ], voting="hard"),
    svm.SVC(random_state=RANDOM_SEED),
    neighbors.KNeighborsClassifier(),
    naive_bayes.GaussianNB(),
    linear_model.LogisticRegression(max_iter=150, random_state=RANDOM_SEED),
    neural_network.MLPClassifier(random_state=RANDOM_SEED),
    discriminant_analysis.LinearDiscriminantAnalysis(),
    gaussian_process.GaussianProcessClassifier(random_state=RANDOM_SEED),
]

# Measure the performance of all models on the training and validation data.
measurements_df = measure_models(all_models, num_train_x.values, train_y, num_valid_x.values, valid_y)
measurements_df

/home/den/anaconda3/envs/ds12/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning:

lbfgs failed to converge after 150 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=150).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression

/home/den/anaconda3/envs/ds12/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning:

lbfgs failed to converge after 150 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=150).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refe

,Model,Accuracy,Precision,Recall,F1 Score,Validation Accuracy,Validation Precision,Validation Recall,Validation F1 Score
0,MLPClassifier,0.812874,0.752941,0.755906,0.754420,0.798206,0.736264,0.761364,0.748603
1,RandomForestClassifier,0.989521,0.995984,0.976378,0.986083,0.793722,0.728261,0.761364,0.744444
2,GradientBoostingClassifier,0.911677,0.922078,0.838583,0.878351,0.789238,0.766234,0.670455,0.715152
3,VotingClassifier,0.913174,0.937500,0.826772,0.878661,0.789238,0.753086,0.693182,0.721893
4,AdaBoostClassifier,0.989521,0.992032,0.980315,0.986139,0.780269,0.705263,0.761364,0.732240
5,HistGradientBoostingClassifier,0.949102,0.958333,0.905512,0.931174,0.780269,0.724138,0.715909,0.720000
6,DecisionTreeClassifier,0.989521,0.995984,0.976378,0.986083,0.775785,0.720930,0.704545,0.712644
7,ExtraTreesClassifier,0.989521,0.995984,0.976378,0.986083,0.775785,0.693878,0.772727,0.731183
8,LogisticRegression,0.814371,0.787611,0.700787,0.741667,0.766816,0.714286,0.681818,0.697674
9,LinearDiscriminantAnalysis,0.806886,0.765957,0.708661,0.736196,0.766816,0.704545,0.704545,0.704545


### i) **Visualization of Model Performance**

In [213]:
plot_models_performance(
measurements_df, metric="Validation Accuracy"
)

In [214]:
plot_models_performance(
measurements_df, metric="Validation F1 Score"
)

In [215]:
plot_models_performance(
    measurements_df, metric="Validation Precision"
)

In [216]:
plot_models_performance(
    measurements_df, metric="Validation Recall"
)

<h1 style="font-size: 15px; text-align: center;">Bless Be LORD GOD Of hosts</h1>
<p style="font-size: 15px; text-align: center;">copyright&copy;2025</p>

<a id="bottom"></a>